File: python/mirage/core/jaeger/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.core.hierarchy.scope import RouteMatch
from mirage.core.hierarchy.stat import make_stat
from mirage.core.jaeger.readdir import readdir, service_guard
from mirage.core.jaeger.scope import detect_scope


def _service_extra(match: RouteMatch) -> dict[str, str]:
    return {"service": match.captures["service"]}


def _trace_extra(match: RouteMatch) -> dict[str, str]:
    return {"trace_id": match.captures["trace_id"]}


stat = make_stat(
    detect_scope,
    readdir,
    guards={
        "service": service_guard,
        "traces": service_guard,
    },
    extras={
        "service": _service_extra,
        "trace": _trace_extra,
    },
)

File: typescript/packages/core/src/core/jaeger/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { RouteMatch } from '../hierarchy/scope.ts'
import { makeStat } from '../hierarchy/stat.ts'
import { readdir, serviceGuard } from './readdir.ts'
import { detectScope } from './scope.ts'

function serviceExtra(match: RouteMatch): Record<string, string> {
  return { service: match.captures.service ?? '' }
}

function traceExtra(match: RouteMatch): Record<string, string> {
  return { trace_id: match.captures.trace_id ?? '' }
}

export const stat = makeStat(detectScope, readdir, {
  guards: {
    service: serviceGuard,
    traces: serviceGuard,
  },
  extras: {
    service: serviceExtra,
    trace: traceExtra,
  },
})
